circuitbreaker/circuit_breaker.go

```diff
@@ -0,0 +1,413 @@
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	if cb.state == StateOpen {
+		cb.state = StateHalfOpen
+		cb.lastStateChange = cb.now()
+
+		// Reset counters
+		atomic.StoreInt64(&cb.failureCount, 0)
+		atomic.StoreInt64(&cb.successCount, 0)
+
+		// Empty the semaphore channel
+		select {
+		case <-cb.halfOpenSemaphore:
+		default:
+		}
+	}
+}
+
+// transitionToClosed changes state from half-open to closed
+func (cb *CircuitBreaker) transitionToClosed() {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	if cb.state == StateHalfOpen {
+		cb.state = StateClosed
+		cb.lastStateChange = cb.now()
+
+		// Reset counters
+		atomic.StoreInt64(&cb.failureCount, 0)
+		atomic.StoreInt64(&cb.successCount, 0)
+	}
+}
+
+// AllowRequest determines if a request is allowed based on circuit state
+func (cb *CircuitBreaker) AllowRequest() (bool, State) {
+	atomic.AddInt64(&cb.totalRequests, 1)
+
+	cb.mutex.RLock()
+	state := cb.state
+	cb.mutex.RUnlock()
+
+	switch state {
+	case StateOpen:
+		atomic.AddInt64(&cb.rejectedRequests, 1)
+		return false, state
+	case StateHalfOpen:
+		select {
+		case cb.halfOpenSemaphore <- struct{}{}:
+			return true, state
+		default:
+			atomic.AddInt64(&cb.rejectedRequests, 1)
+			return false, state
+		}
+	default: // StateClosed
+		return true, state
+	}
+}
+
+// ReleaseSemaphore releases a slot in the half-open semaphore
+func (cb *CircuitBreaker) ReleaseSemaphore() {
+	select {
+	case <-cb.halfOpenSemaphore:
+	default:
+	}
+}
+
+// ReportSuccess increments success count and closes circuit if threshold met
+func (cb *CircuitBreaker) ReportSuccess() {
+	cb.mutex.RLock()
+	currentState := cb.state
+	cb.mutex.RUnlock()
+
+	if currentState == StateHalfOpen {
+		newSuccessCount := atomic.AddInt64(&cb.successCount, 1)
+		if int(newSuccessCount) >= cb.successThreshold {
+			cb.transitionToClosed()
+		}
+	}
+}
+
+// ReportFailure increments failure count and opens circuit if threshold met
+func (cb *CircuitBreaker) ReportFailure() {
+	cb.mutex.RLock()
+	currentState := cb.state
+	cb.mutex.RUnlock()
+
+	switch currentState {
+	case StateHalfOpen:
+		// In half-open, a single failure trips the circuit
+		cb.transitionToOpen()
+	case StateClosed:
+		newFailureCount := atomic.AddInt64(&cb.failureCount, 1)
+		if int(newFailureCount) >= cb.failureThreshold {
+			cb.transitionToOpen()
+		}
+	}
+}
+
+// Metrics returns basic metrics about the circuit breaker
+func (cb *CircuitBreaker) Metrics() map[string]interface{} {
+	return map[string]interface{}{
+		"state":            cb.GetState(),
+		"failures":         atomic.LoadInt64(&cb.failureCount),
+		"successes":        atomic.LoadInt64(&cb.successCount),
+		"totalRequests":    atomic.LoadInt64(&cb.totalRequests),
+		"rejectedRequests": atomic.LoadInt64(&cb.rejectedRequests),
+	}
+}
+
+// GetStateStats returns detailed statistics about the circuit breaker
+func (cb *CircuitBreaker) GetStateStats() map[string]interface{} {
+	state := cb.GetState()
+
+	return map[string]interface{}{
```

Comment:
these stats should have their own struct with field and their types. returning map of interfaces looks like something that belongs to javascript world.